### PR TITLE
Update JAX to 0.4.13

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context**: Update to JAX 0.4.13. [`llvm::Optional` is being deprecated ](https://discourse.llvm.org/t/deprecating-llvm-optional-x-hasvalue-getvalue-getvalueor/63716). Inlining has to be registered.

ArithConstantOp uses InferResultType but it is not shipped with JAX. This means that we either patch the auto-generated file or use a different operation. I opted for using a different operation here.

**Description of the Change:**
1. Use `std::optional` instead of `LLVM::Optional`
2. Register inlining extension
3. Fix patch
4. Avoid using ArithConstantOp
5. Style

### CODE REVIEW

**Comment by erick-xanadu** [comment]
Action for building the wheel (in progress, will take several hours): https://github.com/PennyLaneAI/catalyst/actions/runs/5465642638

File: mlir/tools/quantum-opt/quantum-opt.cpp [hunk 1]
```diff
@@ -39,6 +40,7 @@ int main(int argc, char **argv)
 
     mlir::DialectRegistry registry;
     mlir::registerAllDialects(registry);
+    mlir::func::registerAllExtensions(registry);
```
Comment (maliasadi) [inline mlir/tools/quantum-opt/quantum-opt.cpp:43]: For my knowledge, is this a strong requirement from using the newer version of LLVM/MLIR?

**Review by maliasadi — COMMENTED** [review]
Thanks @erick-xanadu! Nothing much from my side. We can merge this after checking the wheels. :tada:

File: .github/workflows/build-wheel-linux-x86_64.yaml [hunk 2]
```diff
@@ -70,7 +70,7 @@ jobs:
               -DLLVM_ENABLE_ASSERTIONS=ON \
               -DLLVM_INSTALL_UTILS=ON
 
-        cmake --build llvm-build --target lld check-mlir -j$(nproc)
+        LIT_FILTER_OUT="Bytecode" cmake --build llvm-build --target lld check-mlir -j$(nproc)
```
Comment (maliasadi) [inline .github/workflows/build-wheel-linux-x86_64.yaml:73]: If everything passes, let's add a note here and there. 

We may need to update `mlir/Makefile` as well.

File: .dep-versions [hunk 3]
```diff
@@ -1,3 +1,3 @@
-jax=0.4.10
-mhlo=14675b981264da44b0bb18217559d8cbbb12d910
-llvm=e13d1b5227a77bb7becfd4c49a60720ecc33f870
+jax=0.4.13
+mhlo=72c2cd05cd9c2f1162890c507b1856afa4b4d699
+llvm=73a0ae021ec6112911b98025055b8a2e881b1376
```
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply .dep-versions:3]: See here: https://github.com/PennyLaneAI/catalyst/blob/main/.github/workflows/set_dep_versions.py

File: doc/changelog.md [hunk 4]
```diff
@@ -222,6 +222,9 @@
   This improves IR readability.
   [#165](https://github.com/PennyLaneAI/catalyst/pull/165)
 
+* Update JAX to `v0.4.13`.
```
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply doc/changelog.md:225]:
```suggestion
* Update JAX to ``v0.4.13``.
```

File: frontend/catalyst/jax_primitives.py [hunk 5]
```diff
@@ -1426,8 +1426,19 @@ def _qfor_lowering(
...
+        zero_attr = ir.DenseIntElementsAttr.get(zero_np)
+        one_attr = ir.DenseIntElementsAttr.get(one_np)
+        zero_tensor = StableHLOConstantOp(zero_attr)
+        one_tensor = StableHLOConstantOp(one_attr)
+        ctx = jax_ctx.module_context.context
+        i64_type = ir.IntegerType.get_signless(64, ctx)
+        zero_i64 = TensorExtractOp(i64_type, zero_tensor, []).result
+        one_i64 = TensorExtractOp(i64_type, one_tensor, []).result
+        zero = IndexCastOp(ir.IndexType.get(), zero_i64).result
+        one = IndexCastOp(ir.IndexType.get(), one_i64).result
```
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/jax_primitives.py:1440]: > instead of patching?

We are not patching here. I think what you are saying is suggesting a third approach, one which is not patching, not the sequence of instructions above, but instead adding a python class. Is this correct?

File: mlir/tools/quantum-opt/quantum-opt.cpp (hunk 1 above)
Comment (erick-xanadu) [reply mlir/tools/quantum-opt/quantum-opt.cpp:43]: Yes, see [here](https://github.com/llvm/llvm-project/commit/a5ef51d786f241cdd9d1bec62fbc2bf89a766b7d).

File: frontend/catalyst/jax_primitives.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (maliasadi) [reply frontend/catalyst/jax_primitives.py:1440]: Yep, that's correct! I was forming my thoughts for a third approach to get rid of patches :)

[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/jax_primitives.py:1440]: The answer is yes, but I think it is not what you want. Creating a custom constructor / class is the preferred way of doing this. The [MLIR Python Bindings documentation](https://mlir.llvm.org/docs/Bindings/Python/#extending-generated-op-classes) agrees with you. However, they have some machinery that automatically looks for a file with the extensions and will dynamically create a subclass that inherits from both (the default implementation and the extension). This makes it so that all extensions to the dialects are written in the `_ext.py` file.

However, this extension mechanism is intended for custom dialects. Dialects already available in MLIR may come with their own `_ext.py` file. The arithmetic dialect does so. And modifying it is equivalent to us patching it.

I've tried not using the `_ext.py` file to make a constructor, but I think the meta-programming gets in the way.

File: doc/changelog.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply doc/changelog.md:225]: I'm waiting until the wheel passes before adding more changes.

**Comment by erick-xanadu** [comment]
After filtering out the test building wheels succeed. See [here](https://github.com/PennyLaneAI/catalyst/actions/runs/5467964871).

File: .github/workflows/build-wheel-linux-x86_64.yaml (hunk 2 above)
Comment (erick-xanadu) [reply .github/workflows/build-wheel-linux-x86_64.yaml:75]: https://github.com/PennyLaneAI/catalyst/pull/185/commits/f4c28d6719f87d61bd9007a8e96883eb025b3d56

File: frontend/catalyst/jax_primitives.py [hunk 6]
```diff
@@ -1426,8 +1426,19 @@ def _qfor_lowering(
         loop_operands.append(p)
 
     if apply_reverse_transform:
-        zero = ArithConstantOp(ir.IndexType.get(), 0)
-        one = ArithConstantOp(ir.IndexType.get(), 1)
+        zero_np = np.array(0)
```
Comment (rmoyard) [inline frontend/catalyst/jax_primitives.py:1429]: Is it something that will be reverted in a potential future Jax release (support for InferResultType for Arith)? If yes I would add a comment for an easier revisit. If no, ignore this comment.

File: mlir/patches/use-ir-from-jax.patch [hunk 7]
```diff
@@ -1,103 +1,136 @@
 --- ir.py
 +++ ir.py
-@@ -2,95 +2,5 @@
- #  See https://llvm.org/LICENSE.txt for license information.
- #  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
- 
+@@ -1,131 +1,2 @@
+-#  Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
```
Comment (rmoyard) [inline mlir/patches/use-ir-from-jax.patch:4]: Just to better understand, is it a copy paste of the ir generated by the Jax project?

**Review by rmoyard — COMMENTED** [review]
I have left a review to learn more about patches and the latest llvm changes, thanks it looks good :+1:

File: frontend/catalyst/jax_primitives.py (hunk 6 above)
Comment (erick-xanadu) [reply frontend/catalyst/jax_primitives.py:1429]: I don't believe JAX has plans to include the arithmetic dialect in their release. Instead they will continue using StableHLO and only StableHLO.

File: mlir/patches/use-ir-from-jax.patch (hunk 7 above)
Comment (erick-xanadu) [reply mlir/patches/use-ir-from-jax.patch:4]: This is an autogenerated file when creating the python bindings. However it conflicts with the python bindings provided by JAX. Here I am removing the attribute builders that need to be registered. JAX includes these as well and re-registering them causes an error.

File: frontend/catalyst/jax_primitives.py (hunk 6 above)
Comment (rmoyard) [reply frontend/catalyst/jax_primitives.py:1429]: Thanks!

File: mlir/patches/use-ir-from-jax.patch (hunk 7 above)
Comment (rmoyard) [reply mlir/patches/use-ir-from-jax.patch:4]: Thank you for the details

**Review by rmoyard — APPROVED** [review]
Thank you @erick-xanadu!

File: frontend/catalyst/jax_primitives.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/jax_primitives.py:1440]: There might be a different way. I'm investigating it further today.

**Review by maliasadi — APPROVED** [review]

File: frontend/catalyst/jax_primitives.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/jax_primitives.py:1440]: Just for completeness: 

I think I found a way to get rid of all of these patches. Again, this will mean that we will get rid of the python bindings for the quantum and the gradient dialect in the project. Essentially, we will no longer need the python bindings for any dialects as we will use the ones provided by JAX and or create operations as python objects using the generic interface. (See below)

The way to do so is to use the generic operation constructor for all dialects/operations which are not included by JAX. See here (https://mlir.llvm.org/docs/Bindings/Python/#operations-regions-and-blocks-1)

There is one note though and that this is discouraged. 

> Operations can also be constructed using the generic class and based on the
canonical string name of the operation using Operation.create. It accepts the
operation name as string, which must exactly match the canonical name of the
operation in C++ or ODS, followed by the same argument list as the default
constructor for OpView. This form is discouraged from use and is intended
for generic operation processing.

I personally think this is not bad for our use case, but the notice worries me a bit.
